test/src/helper/chai-similar.ts

```diff
@@ -0,0 +1,191 @@
+import * as _ from "lodash";
+import { isNullOrUndefined } from "util";
+
+export const $else = Symbol("else");
+
+export function $anything() {
+    return true;
+}
+$anything.toString = () => "$anything";
+
+export function $anyOf(values: any[]) {
+    function pred(actual: any) {
+        for (const expected of values) {
+            if (deepCompare(actual, expected) === true) {
+                return true;
+            }
+        }
+        return false;
+    }
+    pred.toString = function toString() {
+        return "$anyOf(" + JSON.stringify(normalize(values)) + ")";
+    };
+    return pred;
+}
+
+export function $allOf(values: any[]) {
+    function pred(actual: any) {
+        for (const expected of values) {
+            if (deepCompare(actual, expected) === false) {
+                return false;
+            }
+        }
+        return true;
+    }
+    pred.toString = function toString() {
+        return "$allOf(" + JSON.stringify(normalize(values)) + ")";
+    };
+    return pred;
+}
+
+export function $containsWord(word: string) {
+    const escape = /([()])/g;
+    let wordPattern: string = word;
+    if (escape.test(word)) {
+        wordPattern = word.replace(escape, "\\$1");
+    }
+    const regex = new RegExp("\\b" + wordPattern + "\\b");
+    function pred(actual: any) {
+        return regex.test(actual);
+    }
+    pred.toString = function toString() {
+        return "$containsWord(" + word + ")";
+    };
+    return pred;
+}
+
+export function similar(chai: any, utils: any): void {
+    function assertSimilar(this: any, expected: any, msg: boolean) {
+        if (msg) {
+            utils.flag(this, "message", msg);
+        }
+        const actual = utils.flag(this, "object");
+        const result = deepCompare(actual, expected);
+
+        this.assert(
+            result,
+            "not similar",
+            "similar",
+            normalize(expected),
+            normalize(actual),
+            true
+        );
+    }
+
+    utils.addMethod(chai.Assertion.prototype, "similarTo", assertSimilar);
+}
+
+declare global {
+    namespace Chai {
+        type Similar = (value: any, message?: string) => Assertion;
+
+        interface Assertion {
+            similarTo: Similar;
+        }
+    }
+}
+
+function deepCompare(actual: any, expected: any): boolean {
```

Comment:
Is it different from `to.deep.equal` of Chai?

Reply:
chai `to.deep.equal` simply compares the reference even if it is a function.
However, deepCompare uses `expected` as a predicate function of `actual` when it is a function. It enables to customize compare function.